community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -0,0 +1,481 @@
+            cursor.checkAndClearCursorException();
+            if ( cursor.checkAndClearBoundsFlag() )
+            {
+                // Something's wrong, get a new fresh rootId and go from the top again
+                cursor.next( localRootId = rootId );
+                continue;
+            }
+            if ( isInternal )
+            {
+                // TODO perhaps detect being too far to the left and so follow right siblings,
+                // to reduce unnecessary seek in leafs (concurrent splits would be the cause)
+
+                PointerChecking.checkChildPointer( childId );
+
+                if ( !cursor.next( childId ) )
+                {
+                    throw new IllegalStateException( "Couldn't go to child " + childId );
+                }
+            }
+        }
+        while ( isInternal && keyCount > 0 );
+
+        // Returns cursor which is now initiated with left-most leaf node for the specified range
+        return new SeekCursor<>( cursor, key, value, bTreeNode, fromInclusive, toExclusive, layout,
+                stableGeneration, unstableGeneration, pos, keyCount );
+    }
+
+    @Override
+    public long acquireNewId()
+    {
+        lastId++;
+        return lastId;
+    }
+
+    // Utility method
+    public void printTree() throws IOException
+    {
+        try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK ) )
+        {
+            cursor.next();
+            TreePrinter.printTree( cursor, bTreeNode, layout, stableGeneration, unstableGeneration, System.out );
+        }
+    }
+
+    // Utility method
+    public boolean consistencyCheck() throws IOException
+    {
+        try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_READ_LOCK ) )
+        {
+            cursor.next();
+            return new ConsistencyChecker<>( bTreeNode, layout, stableGeneration, unstableGeneration )
+                    .check( cursor );
+        }
+    }
+
+    @Override
+    public void flush() throws IOException
+    {
+        try ( PageCursor cursor = openMetaPageCursor( pagedFile ) )
+        {
+            cursor.putLong( 4, rootId );
+            cursor.putLong( 12, lastId );
+            // generations should be incremented as part of flush, but this functionality doesn't exist yet.
+        }
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        flush();
+        pagedFile.close();
+    }
+
+    /**
+     * @return the single {@link IndexWriter} for this index. The returned writer must be
+     * {@link IndexWriter#close() closed} before another caller can acquire this writer.
+     * @throws IllegalStateException for calls made between a successful call to this method and closing the
+     * returned writer.
+     */
+    @Override
+    public IndexWriter<KEY,VALUE> writer( IndexWriter.Options options ) throws IOException
+    {
+        SingleIndexWriter result = this.writer.getAndSet( null );
```

Comment:
This could be a nice thing for the future if interaction with neo changes and we can't rely on WorkSync. For now, this implementation is bare minimum of what is needed.